From c82720420b4eb6299ad1be6d51e3d07fcd5e6df5 Mon Sep 17 00:00:00 2001
From: Ariadne Conill <ariadne@dereferenced.org>
Date: Fri, 14 Feb 2020 16:02:43 +0000
Subject: [PATCH 1/2] Add support for starting services in a specified VRF.

The venerable iproute2 utility has recently introduced support
for executing programs in specific VRFs which are virtualized
routing tables.  These are typically used to isolate different
networking planes from each other, for security or flexibility
reasons.

Services which use the normal supervisor/start-stop-daemon
pattern can be configured by setting the vrf variable in the
/etc/conf.d tree for the service.

This allows for things like configuring the sshd service to
run in a management VRF, which is useful for high assurance
environments where the management plane is intended to be
isolated.

Signed-off-by: Ariadne Conill <ariadne@dereferenced.org>
---
 sh/openrc-run.sh.in     | 6 ++++++
 sh/runit.sh             | 2 +-
 sh/start-stop-daemon.sh | 2 +-
 sh/supervise-daemon.sh  | 2 +-
 4 files changed, 9 insertions(+), 3 deletions(-)

diff --git a/sh/openrc-run.sh.in b/sh/openrc-run.sh.in
index 11f66a27..f50b1624 100644
--- a/sh/openrc-run.sh.in
+++ b/sh/openrc-run.sh.in
@@ -260,6 +260,12 @@ if ! sourcex -e "$_conf_d/$RC_SVCNAME.$RC_RUNLEVEL"; then
 fi
 unset _conf_d
 
+# If we are configured to run in a VRF, provide a hint for that
+RC_VRF_EXEC=""
+if [ -n "$vrf" ]; then
+	RC_VRF_EXEC="ip vrf exec $vrf"
+fi
+
 if yesno "$RC_USER_SERVICES" && [ "$_usr_conf/init.d" != "${RC_SERVICE%/*}" ]; then
 	if ! sourcex -e "$_usr_conf/conf.d/$RC_SVCNAME.$RC_RUNLEVEL"; then
 		sourcex -e "$_usr_conf/conf.d/$RC_SVCNAME"
diff --git a/sh/runit.sh b/sh/runit.sh
index d4e37f34..91f29cb6 100644
--- a/sh/runit.sh
+++ b/sh/runit.sh
@@ -23,7 +23,7 @@ runit_start()
 	local i=0 retval=1
 	# it can take upto 5 seconds for runsv to start
 	while [ $i -lt 6 ] ; do
-		if sv start "${service_link}" > /dev/null 2>&1; then
+		if ${RC_VRF_EXEC} sv start "${service_link}" > /dev/null 2>&1; then
 			retval=0
 			break
 		fi
diff --git a/sh/s6.sh b/sh/s6.sh
index 5a63fca0..58201d6b 100644
--- a/sh/s6.sh
+++ b/sh/s6.sh
@@ -137,7 +137,7 @@ _s6_servicedir_create() {
 		if test -n "$command_user" ; then
 			echo "s6-setuidgid \"$command_user\""
 		fi
-		echo "sh -c \"exec $command $command_args $command_args_foreground\""
+		echo "sh -c \"exec ${RC_VRF_EXEC} $command $command_args $command_args_foreground\""
 	} > "$dir/run"
 	chmod 0755 "$dir/run"
 
diff --git a/sh/start-stop-daemon.sh b/sh/start-stop-daemon.sh
index b8706128..d5dc911c 100644
--- a/sh/start-stop-daemon.sh
+++ b/sh/start-stop-daemon.sh
@@ -46,7 +46,7 @@ ssd_start()
 	#the eval call is necessary for cases like:
 	# command_args="this \"is a\" test"
 	# to work properly.
-	eval start-stop-daemon --start \
+	eval ${RC_VRF_EXEC} start-stop-daemon --start \
 		--exec $command \
 		${chroot:+--chroot} $chroot \
 		${directory:+--chdir} $directory \
diff --git a/sh/supervise-daemon.sh b/sh/supervise-daemon.sh
index f561e84d..30273979 100644
--- a/sh/supervise-daemon.sh
+++ b/sh/supervise-daemon.sh
@@ -29,7 +29,7 @@ supervise_start()
 	# The eval call is necessary for cases like:
 	# command_args="this \"is a\" test"
 	# to work properly.
-	eval supervise-daemon "${RC_SVCNAME}" --start \
+	eval ${RC_VRF_EXEC} supervise-daemon "${RC_SVCNAME}" --start \
 		${retry:+--retry} $retry \
 		${directory:+--chdir} $directory  \
 		${chroot:+--chroot} $chroot \
-- 
2.47.3

